### DISCUSSION TITLE
How to stop msi installer from being able to install twice?

### DISCUSSION THREAD

**Discussion by Fiinleyyy** [oleg-shilo/wixsharp#1569] [Q&A]
I was so sure that in the past I wasn't able to install the same msi installer twice. But right now if I install the first time everything works as expected. But as soon as try to install the second time with the product already being installed the wizard won't open but instead will install the product again but in silent mode. 

That behaviour is very weird - is the a way to enable upgrades but disable multiple installations of the same product?

**Comment by Ivanco21** [comment]
I have a simular question.
If try to install twice, installer not re install version (not copy files for my app) and not show error.
I want a warning like “this version is already installed” to be displayed.
Tried to realize this behaviour (code bellow), but it not works for me.

How corretly check installed version, and show ActionResult.UserExit for this case?

```
project.UIInitialized += DoBeforeInstall;

private static void DoBeforeInstall(SetupEventArgs e)
{
    Version thisVersion = e.Session.QueryProductVersion();
    string[] productCodes = AppSearch.GetProductCode("My app");
    IEnumerable<string> installedVersionsStr = productCodes.Select(pCode => AppSearch.GetProductVersionString(pCode))
                                                        .Where(v => v != null);
    IEnumerable<Version> installedVersions = installedVersionsStr.Select(v => new Version(v));

    if (installedVersions.Contains(thisVersion))
    {
        e.ManagedUI.Shell.ErrorDetected = true;
        e.ManagedUI.Shell.CustomErrorDescription = "errMsg";
        e.Result = ActionResult.UserExit;
    }
}
```

**Comment by oleg-shilo** [comment]
@Fiinleyyy, by default upgrading is disabled. If you want to implement Major Upgrade concept then you need to set `project.MajorUpgrade` property as in the [MajorUpgrade](https://github.com/oleg-shilo/wixsharp/blob/master/Source/src/WixSharp.Samples/Wix%23%20Samples/MajorUpgrade/setup.cs) sample. IE:

```C#
project.MajorUpgrade = new MajorUpgrade
{
    Schedule = UpgradeSchedule.afterInstallInitialize,
    DowngradeErrorMessage = "A later version of [ProductName] is already installed. Setup will now exit."
};
```

If you do not set `project.MajorUpgrade` then msi will always abort the installation if the product is already installed. And if you try to run the same msi twice the second time it will offer you modify/repair/remove options unless you suppress the UI deliberately. 

I just retested the simplest sample [Install_Files](https://github.com/oleg-shilo/wixsharp/tree/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Install%20Files) and it does work as expected. 

I suggest you establish the baseline of a working solution and then see what difference in your code is causing the problem. You can do this by simply creating a simple project from the VS template without any modification of the code. Then building and testing it:

![image](https://github.com/user-attachments/assets/71c7fc8b-ed11-4af1-9b3c-f6081d6c1475)
